src/ompl/geometric/planners/rrt/src/RRTstar.cpp

```diff
@@ -573,73 +652,322 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
     }
 }
 
-int ompl::geometric::RRTstar::pruneTree(const base::Cost pruneTreeCost)
+int ompl::geometric::RRTstar::pruneTree(const base::Cost& pruneTreeCost)
 {
-    const int tree_size = nn_->size();
-    pruneScratchSpace_.newTree.reserve(tree_size);
-    pruneScratchSpace_.newTree.clear();
-    pruneScratchSpace_.toBePruned.reserve(tree_size);
-    pruneScratchSpace_.toBePruned.clear();
-    pruneScratchSpace_.candidates.clear();
-    pruneScratchSpace_.candidates.push_back(startMotion_);
-    std::size_t j = 0;
-    while (j != pruneScratchSpace_.candidates.size())
+    // Variable
+    // The percent improvement (expressed as a [0,1] fraction) in cost
+    double fracBetter;
+    // The number pruned
+    int numPruned = 0;
+
+    if (opt_->isFinite(prunedCost_))
     {
-        Motion *candidate = pruneScratchSpace_.candidates[j++];
-        if (opt_->isCostBetterThan(pruneTreeCost, costToGo(candidate)))
-            pruneScratchSpace_.toBePruned.push_back(candidate);
-        else
+        fracBetter = std::abs((pruneTreeCost.value() - prunedCost_.value())/prunedCost_.value());
+    }
+    else
+    {
+        fracBetter = 1.0;
+    }
+
+    if (fracBetter > pruneThreshold_)
+    {
+        // We are only pruning motions if they, AND all descendents, have a estimated cost greater than pruneTreeCost
+        // The easiest way to do this is to find leaves that should be pruned and ascend up their ancestry until a motion is found that is kept.
+        // To avoid making an intermediate copy of the NN structure, we process the tree by descending down from the start(s).
+        // In the first pass, all Motions with a cost below pruneTreeCost, or Motion's with children with costs below pruneTreeCost are added to the replacement NN structure,
+        // while all other Motions are stored as either a 'leaf' or 'chain' Motion. After all the leaves are disconnected and deleted, we check
+        // if any of the the chain Motions are now leaves, and repeat that process until done.
+        // This avoids (1) copying the NN structure into an intermediate variable and (2) the use of the expensive NN::remove() method.
+
+        // Variable
+        // The queue of Motions to process:
+        std::list<Motion*> motionQueue;
+        // The list of leaves to prune
+        std::list<Motion*> leavesToPrune;
+        // The list of chain vertices to recheck after pruning
+        std::list<Motion*> chainsToRecheck;
```

Comment:
Ok, let's leave things as you have them for now. Try running ompl_benchmark on some simple problems with the old RRT* and new RRT* code (both with pruning on) to make sure nothing broke and, ideally, that the new version is faster.

As a general rule of thumb, it's almost always better to use std::vector over std::list *unless* you frequently need to remove elements. Instead of mentioning "memory coherence" I should said that std::vector's are more cache friendly. It may not matter much if you still have to deference Motion*'s, so I am willing to leave it the way you have it, provided it doesn't make RRT* significantly slower.